sw/airborne/modules/guidance/gvf/gvf.c

```diff
@@ -196,38 +227,104 @@ bool gvf_line_wp1_wp2(uint8_t wp1, uint8_t wp2)
   float x2 = waypoints[wp2].x;
   float y2 = waypoints[wp2].y;
 
-  float zx = x1 - x2;
-  float zy = y1 - y2;
+  float zx = x2 - x1;
+  float zy = y2 - y1;
 
-  float alpha = atanf(zy / zx);
+  float heading = atanf(zx / zy);
+
+  gvf_line_XY_heading(x1, y1, heading);
 
-  gvf_line(x1, y1, alpha);
+  horizontal_mode = HORIZONTAL_MODE_ROUTE;
+  gvf_segment.seg = 1;
+  gvf_segment.x1 = x1;
+  gvf_segment.y1 = y1;
+  gvf_segment.x2 = x2;
+  gvf_segment.y2 = y2;
 
   return true;
 }
 
-bool gvf_line_wp_heading(uint8_t wp, float alpha)
+int out_of_segment_area(float x1, float y1, float x2, float y2, float d1, float d2)
 {
-  alpha = alpha * M_PI / 180;
+  struct EnuCoor_f *p = stateGetPositionEnu_f();
+  float px = p->x - x1;
+  float py = p->y - y1;
+
+  float zx = x2 - x1;
+  float zy = y2 - y1;
+  float alpha = atan2f(zy, zx);
+
+  float cosa = cosf(-alpha);
+  float sina = sinf(-alpha);
+
+  float pxr = px * cosa - py * sina;
+  float zxr = zx * cosa - zy * sina;
+
+  int s = 0;
+
+  if (pxr < -d1) {
+    s = 1;
+  } else if (pxr > (zxr + d2)) {
+    s = -1;
+  }
+
+  if (zy < 0) {
+    s *= -1;
+  }
+
+  return s;
+}
+
+bool gvf_segment_XY1_XY2(float x1, float y1, float x2, float y2, float d1, float d2)
+{
+  int s = out_of_segment_area(x1, y1, x2, y2, d1, d2);
+  if (s != 0) {
+    gvf_control.s = s;
+  }
+
+  gvf_line_XY1_XY2(x1, y1, x2, y2);
+  return true;
+}
+
+bool gvf_segment_wp1_wp2(uint8_t wp1, uint8_t wp2, float d1, float d2)
+{
+  float x1 = waypoints[wp1].x;
+  float y1 = waypoints[wp1].y;
+  float x2 = waypoints[wp2].x;
+  float y2 = waypoints[wp2].y;
+
+  int s = out_of_segment_area(x1, y1, x2, y2, d1, d2);
+  if (s != 0) {
+    gvf_control.s = s;
+  }
+
+  gvf_line_XY1_XY2(x1, y1, x2, y2);
+  return true;
+}
```

Comment:
For the segment functions, I have renamed them to segment_loop. For the next update I will add the 'segment function' so it will return false once the segment is finished (I need to think how to implement the logic for this new function).